### PR TITLE
starting metadata sunset

### PR DESCRIPTION
**Pull request by bcoca**
##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

metadata

### CODE REVIEW

File: test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/main.py [hunk 1]
```diff
@@ -889,49 +872,11 @@ def _validate_docs(self):
...
-                        msg='ANSIBLE_METADATA.status must be exactly one of "deprecated" or "removed"'
-                    )
-        else:
-            # We are testing a collection
-            if self.routing and self.routing.get('plugin_routing', {}).get('modules', {}).get(self.name, {}).get('deprecation', {}):
-                # meta/routing.yml says this is deprecated
-                routing_says_deprecated = True
-                deprecated = True
+        # We are testing a collection
+        if self.routing and self.routing.get('plugin_routing', {}).get('modules', {}).get(self.name, {}).get('deprecation', {}):
```
Comment (gundalow) [inline test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/main.py]: Should this be inside a `if self.collection:`?

Comment (bcoca) [reply test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/main.py]: unsure, since routing can possibly handle base deprecations also

File: docs/templates/plugin.rst.j2 [hunk 2]
```diff
@@ -424,19 +424,19 @@ Status
...
 
 {%     endif %}
 
-{%     if metadata.supported_by %}
+{%     if metadata.supported_by|default(False) %}
 {%       set supported_by = support[metadata.supported_by] %}
 - This @{ plugin_type }@ is :ref:`maintained by @{ supported_by }@ <modules_support>`. *[@{ metadata.supported_by }@]*
 
-{%       if metadata.supported_by in ('core', 'network') %}
+{%       if metadata.supported_by|default(False) in ('core', 'network') %}
```
Comment (abadger) [inline docs/templates/plugin.rst.j2]: At the docs meeting yesterday we agreed that metadata should just be removed from the docs altogether.  So just remove the block altogether.

**Comment by thaumos** [comment]
Thanks for updating ansible-test in this as well.

File: docs/templates/plugin.rst.j2 (hunk 2 above)
Comment (bcoca) [reply docs/templates/plugin.rst.j2]: ok, i had understood we were removing from the plugins, but still allowing it's display, but i'll remove this from the template.

**Comment by mattclay** [comment]
@bcoca Is validate-modules supposed to report an error if metadata is found? https://github.com/ansible/community/issues/346#issuecomment-627923457 indicates that it should.

**Comment by bcoca** [comment]
@mattclay depends, i would say 'yes for base' but we are leaving it as 'optional' to appear in modules outside base, since I would not want those to fail and we will still display it if it is present in `ansible-doc`. Sooo optional test we enable in base?

**Comment by mattclay** [comment]
@bcoca Here's another metadata reference that needs to be updated:

https://github.com/ansible/ansible/blob/5bf84007fb60033a3279caa934465d94d049994c/test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/main.py#L1039-L1046

**Comment by bcoca** [comment]
that file does not exist in repo?

**Comment by mattclay** [comment]
@bcoca The `hacking/report.py` tool should not be removed. The functionality that depends on metadata can be removed, but not the entire script.

**Comment by bcoca** [comment]
ah, sorry, thought it was part of metadata editor that called it

File: packaging/release/changelogs/changelog.py [hunk 3]
```diff
@@ -180,14 +180,17 @@ def load_plugins(version, force_reload):
         plugins_data['plugins'] = {}
 
         for plugin_type in C.DOCUMENTABLE_PLUGINS:
-            output = subprocess.check_output([os.path.join(BASE_DIR, 'bin', 'ansible-doc'),
-                                              '--json', '--metadata-dump', '-t', plugin_type])
-            plugins_data['plugins'][plugin_type] = json.loads(to_text(output))
+            list_output = subprocess.check_output([os.path.join(BASE_DIR, 'bin', 'ansible-doc'), '--json', '--list', '-t', plugin_type])
```
Comment (jimi-c) [inline packaging/release/changelogs/changelog.py]: Are these changes related to the metadata sunset? They don't seem to be.

Comment (felixfontein) [reply packaging/release/changelogs/changelog.py]: This change is needed because the `--metadata-dump` of ansible-doc is also nuked (which happens to be unrelated to ANSIBLE_METADATA, and should be removed eventually anyway - that could also be done in a follow-up PR though).

File: lib/ansible/modules/service_facts.py [hunk 4]
```diff
@@ -9,11 +9,7 @@
 __metaclass__ = type
 
 
-ANSIBLE_METADATA = {'metadata_version': '1.1',
-                    'status': ['preview'],
-                    'supported_by': 'core'}
-
-DOCUMENTATION = r'''
+DOCUMENTATION = '''
```
Comment (felixfontein) [inline lib/ansible/modules/service_facts.py]: Why remove the `r` here?

File: lib/ansible/parsing/plugin_docs.py [hunk 5]
```diff
@@ -54,33 +56,16 @@ def read_docstring(filename, verbose=True, ignore_errors=True):
...
-
-            if 'supported_by' not in data['metadata']:
-                data['metadata']['supported_by'] = 'community'
+                            else:
+                                # string should be yaml if already not a dict
+                                data[varkey] = AnsibleLoader(child.value.s, file_name=filename).get_single_data()
 
-            if 'status' not in data['metadata']:
-                data['metadata']['status'] = ['preview']
+                        display.debug('assigned :%s' % varkey)
```
Comment (felixfontein) [inline lib/ansible/parsing/plugin_docs.py]:
```suggestion
                        display.debug('assigned: %s' % varkey)
```

File: test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/main.py [hunk 6]
```diff
@@ -2139,46 +2074,14 @@ def validate(self):
...
-                            strict_ansible_version = self.StrictVersion('.'.join(ansible_version.split('.')[:2]))
-                            end_of_deprecation_should_be_removed_only = strict_ansible_version >= removed_in
-                        elif self.collection_version:
-                            strict_ansible_version = self.collection_version
-                            end_of_deprecation_should_be_removed_only = strict_ansible_version >= removed_in
-                        else:
-                            end_of_deprecation_should_be_removed_only = False
+            if docs and 'deprecated' in docs and docs['deprecated'] is not None:
+                try:
+                    removed_in = StrictVersion(str(docs.get('deprecated')['removed_in']))
```
Comment (felixfontein) [inline test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/main.py]: The old code checked both `removed_in` and `removed_at_date`. Also it did handle tagged versions, and understood collection versions vs. Ansible versions.

**Review by felixfontein — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/service_facts.py (hunk 4 above)
Comment (bcoca) [reply lib/ansible/modules/service_facts.py]: not on purpose .... but i've rebased 30x sooo, probably merged from somewhere

Comment (felixfontein) [reply lib/ansible/modules/service_facts.py]:
```suggestion
DOCUMENTATION = r'''
```

File: test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/main.py (hunk 6 above)
Comment (felixfontein) [reply test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/main.py]: I think this should simply be
```.py
            if docs and 'deprecated' in docs and docs['deprecated'] is not None:
                end_of_deprecation_should_be_removed_only = False
                if 'removed_at_date' in docs['deprecated']:
                    try:
                        removed_at_date = docs['deprecated']['removed_at_date']
                        if parse_isodate(removed_at_date) < datetime.date.today():
                            msg = "Module's deprecated.removed_at_date date '%s' is before today" % removed_at_date
                            self.reporter.error(
                                path=self.object_path,
                                code='deprecated-date',
                                msg=msg,
                            )
                    except ValueError:
                        # Already checked during schema validation
                        pass
                if 'removed_in' in docs['deprecated']:
                    try:
                        collection_name, version = self._split_tagged_version(docs['deprecated']['removed_in'])
                        if collection_name != self.collection_name:
                            self.reporter.error(
                                path=self.object_path,
                                code='invalid-module-deprecation-source',
                                msg=('The deprecation version for a module must be added in this collection')
                            )
                            # Treat the module as not to be removed:
                            raise ValueError('')
                        removed_in = self.StrictVersion(str(version))
                    except ValueError:
                        end_of_deprecation_should_be_removed_only = False
                    else:
                        if not self.collection:
                            strict_ansible_version = self.StrictVersion('.'.join(ansible_version.split('.')[:2]))
                            end_of_deprecation_should_be_removed_only = strict_ansible_version >= removed_in
                        elif self.collection_version:
                            strict_ansible_version = self.collection_version
                            end_of_deprecation_should_be_removed_only = strict_ansible_version >= removed_in
                        else:
                            end_of_deprecation_should_be_removed_only = False
```
i.e. everything that was removed except the first `if` branch on ANSIBLE_METADATA.

File: test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/main.py [hunk 7]
```diff
@@ -2165,11 +2086,14 @@ def validate(self):
                                 code='invalid-module-deprecation-source',
                                 msg=('The deprecation version for a module must be added in this collection')
                             )
-                            # Treat the module as not to be removed:
-                            raise ValueError('')
-                        removed_in = self.StrictVersion(str(version))
+
+                            end_of_deprecation_should_be_removed_only = False
```
Comment (felixfontein) [inline test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/main.py]: The way you've changed this code, it now no longer works: if `end_of_deprecation_should_be_removed_only` is set to` False` here without raising an exception, the `else:` after the `except` will get executed, but that code assumes that `removed_in` has been set. So if this happens, validate-module will now crash.

**Review by felixfontein — CHANGES_REQUESTED** [review]

File: test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/main.py [hunk 8]
```diff
@@ -2165,20 +2085,36 @@ def validate(self):
...
+
+                    except ValueError:
+                        self.reporter.trace(
+                            path=self.object_path,
+                            tracebk=traceback.format_exc()
+                        )
+
+                    if not self.collection:
+                        strict_ansible_version = self.StrictVersion('.'.join(ansible_version.split('.')[:2]))
+                        end_of_deprecation_should_be_removed_only = strict_ansible_version >= removed_in
```
Comment (felixfontein) [inline test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/main.py]: This will crash, since `removed_in` has not been defined when the above `except` goes through.

File: test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/main.py [hunk 9]
```diff
@@ -2139,23 +2074,8 @@ def validate(self):
...
-                            msg = "Module's deprecated.removed_at_date date '%s' is before today" % removed_at_date
-                            self.reporter.error(
-                                path=self.object_path,
-                                code='deprecated-date',
-                                msg=msg,
-                            )
-                    except ValueError:
-                        # Already checked during schema validation
-                        pass
+            if docs.get('deprecated', False):
```
Comment (felixfontein) [inline test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/main.py]: This crashes when `docs` is `None`.

**Review by felixfontein — APPROVED** [review]

File: docs/templates/plugin.rst.j2 [hunk 10]
```diff
@@ -418,35 +418,6 @@ Status
...
-{%       endfor %}
-
-{%     endif %}
-
-{%     if metadata.supported_by %}
-{%       set supported_by = support[metadata.supported_by] %}
-- This @{ plugin_type }@ is :ref:`maintained by @{ supported_by }@ <modules_support>`. *[@{ metadata.supported_by }@]*
-
-{%       if metadata.supported_by in ('core', 'network') %}
-Red Hat Support
```
Comment (jborean93) [inline docs/templates/plugin.rst.j2:440]: Considering all the plugins still in ansible/ansible are supported by us should we just keep this here? Or maybe have it under a conditional var like `supported` that includes this text for "Supported" content that the doc generator can add where necessary.

File: docs/docsite/rst/dev_guide/testing.rst [hunk 11]
```diff
@@ -68,7 +68,6 @@ When Shippable detects an error and it can be linked back to a file that has bee
 
    The test `ansible-test sanity --test validate-modules` failed with the following errors:
```
Comment (relrod) [inline docs/docsite/rst/dev_guide/testing.rst]:
```suggestion
   The test `ansible-test sanity --test validate-modules` failed with the following error:
```

File: docs/docsite/rst/dev_guide/testing.rst [hunk 12]
```diff
@@ -68,7 +68,6 @@ When Shippable detects an error and it can be linked back to a file that has bee
...
    lib/ansible/modules/network/foo/bar.py:0:0: E307 version_added should be 2.4. Currently 2.3
-   lib/ansible/modules/network/foo/bar.py:0:0: E316 ANSIBLE_METADATA.metadata_version: required key not provided @ data['metadata_version']. Got None
 
 From the above example we can see that ``--test pep8`` and ``--test validate-modules`` have identified issues. The commands given allow you to run the same tests locally to ensure you've fixed the issues without having to push your changed to GitHub and wait for Shippable, for example:
```
Comment (relrod) [inline docs/docsite/rst/dev_guide/testing.rst]:
```suggestion
From the above example we can see that ``--test pep8`` and ``--test validate-modules`` have identified an issue. The commands given allow you to run the same tests locally to ensure you've fixed all issues without having to push your changes to GitHub and wait for Shippable, for example:
```

File: lib/ansible/cli/doc.py [hunk 13]
```diff
@@ -286,20 +279,13 @@ def get_plugin_metadata(plugin_type, plugin_name):
...
-            doc, __, __, metadata = get_docstring(filename, fragment_loader, verbose=(context.CLIARGS['verbosity'] > 0),
-                                                  collection_name=collection_name, is_module=(plugin_type == 'module'))
+            doc, __, __, __ = get_docstring(filename, fragment_loader, verbose=(context.CLIARGS['verbosity'] > 0),
+                                            collection_name=collection_name, is_module=(plugin_type == 'module'))
         except Exception:
             display.vvv(traceback.format_exc())
-            raise AnsibleError(
-                "%s %s at %s has a documentation error formatting or is missing documentation." %
-                (plugin_type, plugin_name, filename))
+            raise AnsibleError("%s %s at %s has a documentation error formatting or is missing documentation." % (plugin_type, plugin_name, filename))
```
Comment (relrod) [inline lib/ansible/cli/doc.py]:
```suggestion
            raise AnsibleError("%s %s at %s has a documentation formatting error or is missing documentation." % (plugin_type, plugin_name, filename))
```

File: docs/templates/plugin.rst.j2 (hunk 10 above)
Comment (bcoca) [reply docs/templates/plugin.rst.j2:440]: no, docs wont reflect it at all, its been decided from other teams to just use AH is supported, rest is not.

File: docs/docsite/rst/dev_guide/testing.rst (hunk 11 above)
Comment (bcoca) [reply docs/docsite/rst/dev_guide/testing.rst]: leaving preexisting text w/o change, iirc it does show multiple errors

File: docs/templates/plugin.rst.j2 (hunk 10 above)
Comment (jborean93) [reply docs/templates/plugin.rst.j2:440]: Yep but still it would be nice to have the builtin docs for Ansible plugins to show the supported status.
